azavea/cac-tripplanner-android
# query destinations

## Overview

Loads actual destinations into the home view carousel and displays the distance to each from a dummy origin (City Hall). Stores destinations to local database; requeries API for destinations if more than the allotted cache time has passed (currently 15 minutes).


## Demo

Pixel 2 emulator on API 27:
![nearby_android_distance](https://user-images.githubusercontent.com/960264/36173582-82d359c4-10d7-11e8-8c34-ecbb35ba453d.png)

Nexus One emulator on API 19:
![nearby_locations_android_api_19](https://user-images.githubusercontent.com/960264/36173588-863eaee2-10d7-11e8-9e52-aeb5d661617f.png)


## Notes

Followed the [Architecture Components Guide](https://developer.android.com/topic/libraries/architecture/guide.html) to set up the new tools (recently out of beta) that better support handling life cycle events, especially the interactions between life cycle events, persistent storage, and network requests. The code snippets in the guide, however, leave a lot out, particularly the parts relating to configuring dependency injection. To fill in the gaps, several files came from the [Github browser example project](https://github.com/googlesamples/android-architecture-components/blob/1532c562bb4d25f8469471478195160fd96b5db7/BasicSample/README.md) that appears to be the source for the snippets in the guide; those files have links to the source file they're derived from in comments at the head of the file.

One significant modification from the sample project approach was to change the caching strategy to add a time stamp to the database entries, instead of relying on an in-memory cache of the last request times.

Note that there is now a compilation warning: `Warning:Supported source version 'RELEASE_7' from annotation processor 'android.arch.persistence.room.RoomProcessor' less than -source '1.8'`. This may be safely ignored. Although other Architecture Components explicitly support Java 8 with a separate package version, Room does not yet, although the code examples in the components documentation use lambdas, which are a Java 8 feature.


## Testing instructions
 * Actual destinations should load in carousel
 * Distances to each destination should display in carousel
 * Destinations should be ordered by distance, and limited to the closest eight
 * Rotating, backgrounding, or restarting the app should behave as expected (no errors)
 * App should only re-query for destinations after the cache period has elapsed (15 minutes)
 * Can change `RATE_LIMIT` in `DestinationRepository` to modify cache period for testing

Closes #4 
Closes #6 
Closes #13
